Add TabBar and BaseViewController

yooxdb opened the pull request:
- 기본적인 TabBar(Newsfeed, Profile, Notification) 추가 및 공통 NavigationBar 설정을 위한 BaseViewController
- 각 ViewController에 `private final var profile: ProfileTabView { return view as! View }`와 같은 코드가 있는 이유:

  - view와 관련된 요소는 모두 UIView 파일에 선언하고, 이를 받는 UIViewController에서 바인딩하려고 합니다. 뷰컨의 `override func loadView() { view = View() }` 때문에 `view.menuTable.bind()`와 `profile.menuTable.bind()`은 동일한 코드이긴 하지만 후자가 조금 더 readable할 것 같아서 넣었습니다. 관련해서 의견 있으시다면 리뷰 부탁드립니다.

yooxdb committed:
Add TabBar and BaseViewController

- 기본적인 TabBar(Newsfeed, Profile, Notification) 추가 및 공통 NavigationBar 설정을 위한 BaseViewController
- Controller 폴더 안의 각 `NewsfeedTabViewController`, `ProfileTabViewController`, `NotificationTabViewController`에 `private final var profile: ProfileTabView { return view as! View }`와 같은 코드가 있는 이유:
  - view와 관련된 요소는 모두 UIView 파일에 선언하고, 이를 받는 UIViewController에서 바인딩하려고 합니다. 위의 `override func loadView() { view = View() }` 때문에 `view.menuTable.bind()`와 `profile.menuTable.bind()`은 동일한 코드이긴 하지만 후자가 조금 더 readable할 것 같아서 넣었습니다. 관련해서 의견 있으시다면 리뷰 부탁드립니다.

shp7724 commented on Facebook/Facebook/Controller/NewsfeedTabViewController.swift:
```diff
@@ -5,15 +5,19 @@
 
+    override func loadView() { view = View() }
+    
+    private final var newsfeed: NewsfeedTabView { return view as! View }
```
여기에서 SwiftLint가 `Force Cast Violation: Force casts should be avoided. (force_cast)`라는 에러를 출력하는데요, `view as! View`가 (코드로 뷰를 바인딩할 때) 일반적으로 자주 사용되는 코드인가요? 만약 그렇다면 SwiftLint에서 해당 Rule을 제거하는게 맞을 것 같아서요.

yooxdb replied:
앗! 남겨주신 코멘트를 확인하지 못하고 SwiftLint 통과를 위해 guard let 문으로 수정하여 commit 하였습니다.

shp7724 replied:
넵 확인했습니다. 추후에라도 SwiftLint 규칙이 쓸데없이 빡센거같다 싶으시면 말씀해주세요 ~ @KWSMooBang 님도요!

yooxdb replied:
`view as! View`는 코드로 뷰를 바인딩할 때 쓰인다기보다는 위의 `loadView()`가 보통 xib 없이 view를 불러올 때 사용되는 코드이고, `view as! View` 이 부분은 위의 PR summary에서 언급했던 것처럼 조금 더 readable한 코드를 위함입니다. View가 항상 존재할 것이므로 force casting을 해도 크게 상관은 없다고 생각하지만 조금 더 안전한 코드를 작성한다 생각하고 guard let 문으로 작성해도 될 것 같네요. 콘솔 출력을 통해 확인해보니 중복적으로 호출되거나 하는 경우는 없는 듯합니다.

shp7724 replied:
아하 이해했습니다. 저렇게 해주어야 type inference가 가능해지는군요. 이 [링크](https://swiftrocks.com/writing-cleaner-view-code-by-overriding-loadview)를 보면 아래와 같은 방식으로도 할 수 있는 것 같은데, 뷰컨마다 `loadView`와 `guard let` 호출 안해도 돼서 VC를 깔끔하게 유지할 수 있다는 장점이 있는 것 같습니다.

```swift

class BaseViewController<CustomTabView: UIView>: UIViewController {
    var tabView: CustomTabView {
        guard let view = view as? CustomTabView else { return CustomTabView() }
        return view
    }

    override func loadView() {
        view = CustomTabView()
    }
}

final class NewsfeedTabViewController: BaseViewController<NewsfeedTabView> {
    // just use tabView.menuTable.bind()
}
```
비록 전처럼 각 뷰마다 별도의 이름을 지정해줄수는 없게 되겠지만, 저는 개인적으로 이 방식이 좀 더 마음에 드는 것 같은데 유림님 의견은 어떠신가요?

---

추가적으로 `BaseViewController`도 하위 클래스 네이밍과의 통일성을 위해 `BaseTabViewController`로 개명(?)하는게 어떨까 싶습니다. 추후에 또다른 Base VC를 만들게 될 수도 있으니깐요.

yooxdb replied:
네 저도 이전에 이 링크의 글을 본 적이 있었는데 이 방법대로 하면 확실히 코드가 매우 깔끔해지긴 합니다..! 하지만 이렇게 되면 신홍님께서 언급해주신 것처럼 뷰마다 별도의 이름을 지정해줄 수 없게 되어 코드의 readability에 별 도움은 되지 못하여 아래 코드를 굳이 작성할 필요가 없어보입니다.
```
var tabView: CustomTabView {
    guard let view = view as? CustomTabView else { return CustomTabView() }
    return view
}
```
별도 명칭을 사용하지 않고 원래대로 `view.~~~`를 사용하는 방식으로 가야할 듯싶네요.

----

`NewsfeedTabViewController`은 `NewsfeedTab` + `ViewController` 의 규칙을 생각하고 지은 이름이어서 `BaseViewController`이라 했던 것인데 충분히 헷갈릴 만한 부분이 있는 것 같네요. 수정해두겠습니다! 좋은 코멘트 감사합니다.

shp7724 replied:
엇 `tabView`를 따로 선언해주지 않고 `view.~~~`와 같은 방식으로 하면, 별도로 타입캐스팅을 해주지 않는 한 `UIView`에 정의된 함수만 호출이 가능하지 않나요? 자동완성도 안될거구요. 그래서 `tabView` 변수는 꼭 필요하다고 생각했습니다.

그리고 사실 뷰와 뷰컨이 일대일대응이라, 뷰의 변수명은 다 똑같아도 뷰컨의 이름으로부터 충분히 추론이 가능하기 때문에 저는 크게 거슬리지는 않았던 것 같습니다! 약간 취향의 영역인 것 같네요.

yooxdb replied:
그렇네요 직접 입력을 통해 확인해보니 별도의 `customView` 선언이 없으면 함수 등을 끌어오기가 어려운 것 같네요. 하위의 모든 뷰컨을 위한 하나의 뷰 변수명을 어떻게 지을 것인지가 관건일 것 같은데, 우선은 저도 마땅한 이름이 떠오르지 않아서 `tabView`로 설정해놓겠습니다!

yooxdb committed:
Modify force cast

- view를 `as! SomethingTabView` 로 unwrapping하는 과정에서 test에 걸려서 guard let 문으로 수정하였습니다

yooxdb commented on Facebook/Facebook/Controller/BaseViewController.swift:
```diff
@@ -0,0 +1,44 @@
+import RxSwift
+import RxGesture
+
+class BaseViewController: UIViewController {
```
@shp7724 @KWSMooBang 그러고보니 PR comment에 다는 것을 깜박했는데 슬랙에서 논의했던 BaseViewController의 navigationBar 설정 방법과 조금 다르게 설정해보았습니다. 이 부분 관련해서도 의견 있으시다면 부탁드립니다! ProfileTabVC에서만 editButton을 추가하고 그걸 navigationItem에 append하는 방식은 일관성이 없다고 판단하여 editButton 또한 BaseVC에서 선언하고 `setNavigationBarItems(withEditButton: Bool = false)` 함수를 작성하였습니다. 기본 값은 false로 지정되어 있어 childVC에서 별도로 호출하지 않으면 searchButton만 등록되고, childVC에서 해당 함수를 호출하고 true를 전달할 경우에는 editButton까지 등록되도록 작성하였습니다.

shp7724 replied:
저는 이 방식 좋은 것 같습니다 ~!

KWSMooBang replied:
저도 이 방식이 기존보다 더 좋은 것 같습니다!

yooxdb committed:
Fix BaseTabViewController

- rename BaseViewController -> BaseTabViewController
- 각 ViewController에서 호출했던 loadView()를 BaseTabViewController에서 일괄적으로 수행하도록 수정하였습니다

yooxdb committed:
Fix BaseTabViewController - 1

- DisposeBag 또한 하위의 모든 뷰컨이 공유하도록 BaseTabViewController에 선언하였습니다
- 하위의 뷰컨에 연결된 각 뷰의 요소들을 사용하기 위하여 UIView형 변수 `tabView`를 추가하였습니다
